### ISSUE TITLE
Digit Grouping Anomaly in '-B' 

### ISSUE THREAD

**Issue by the-solipsist** [simonmichael/hledger#633]
Here's a sample file and outputs. 
Expected output: Same digit grouping for -B and -V (i.e., ₹12,00,000.00)
Actual output: -V outputs correctly (₹12,00,000.00), whereas -B outputs incorrectly (₹1,200,000.00).

Do note: If I remove the "commodity" declaration, or the "P" declaration, or change the amount in the "P" directive from ₹1,000.00 to ₹1000.00, the outputs of -V and -B are the same.
```
╰─$ cat sample.ledger 
commodity ₹1,00,00,00,000.00
P 2017-10-28 $ ₹1,000.00

2017/01/01 * Opening balance
    Assets:Receivable:Lease    ₹12,00,000.00
    Equity:Opening Balances

╰─$ hledger -f sample.ledger bal receivable -V
       ₹12,00,000.00  Assets:Receivable:Lease
--------------------

╰─$ hledger -f sample.ledger bal receivable -B
       ₹1,200,000.00  Assets:Receivable:Lease
--------------------
       ₹1,200,000.00
```

**Comment by simonmichael** [comment]
Thanks! I think the `P` amount is overriding the digit group style set by `commodity`, and it probably shouldn't. Writing `P 2017-10-28 $ ₹0,02,181.90` seems to work around it.

**Comment by simonmichael** [comment]
Notes: 
- print and print -V are [correctly](http://hledger.org/manual.html#amounts) using the digit group style set by the commodity directive ("if there is a commodity directive specifying the format, that is used")
- print -B is incorrectly using the digit group style set by the first amount, which in the above example is in the P directive ("otherwise the format is inferred from the first posting amount in that commodity in the journal")

**Comment by simonmichael** [comment]
Code pointers:
- a single-column balance report renders the amount here, https://github.com/simonmichael/hledger/blob/master/hledger/Hledger/Cli/Commands/Balance.hs#L435 . But the bug happens with other commands too. They all lead to...
- hledger-lib/Hledger/Data/Amount showamountquantity calls punctuatenumber to group digits, according to the style which is stored in the Amount https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Amount.hs#L299
- ...when reading the journal, presumably: 
https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Journal.hs#L480
https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Journal.hs#L717
https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Journal.hs#L744
https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Journal.hs#L836
https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Journal.hs#L849

Maybe journalInferCommodityStyles should also check and prioritise the amounts from commodity directives ? https://github.com/simonmichael/hledger/blob/master/hledger-lib/Hledger/Data/Types.hs#L299

Or, market price amounts should not be included in journalAmounts, to ensure they never influence the standard format ? This would be symmetric with their current behaviour of never being influenced by the standard format (http://hledger.org/manual.html#amounts). But this is delicate and may break something else, code and functional test impact analysis needed.

**Comment by ony** [comment]
This actually a duplicate of #509 . After fixing that I get consistent result from both `-V` and `-B`:
```
zsh% hledger -f /tmp/ws/sample.journal bal receivable -V
       ₹12,00,000.00  Assets:Receivable:Lease
--------------------
       ₹12,00,000.00
zsh% hledger -f /tmp/ws/sample.journal bal receivable -B
       ₹12,00,000.00  Assets:Receivable:Lease
--------------------
       ₹12,00,000.00
```

Though if you would miss `commodity` directive you'll get market price (`P` directive) the first amount in your journal and thus being treated as a canonical format for `₹` currency.

**Comment by ony** [comment]
@simonmichael , should we close this issue as a duplicate of #509?

**Comment by simonmichael** [comment]
Sounds good, please do.

**Comment by ony** [comment]
Duplicate of #509 and fixed by #666. Closing.
